examples_tests/06.MeshLoaders/main.cpp

```diff
@@ -51,22 +51,50 @@ int main()
-    //you might want to IAsset::clone() this DS in order to have exact copy but not the same object. However in this case won't be doing it.
-    //(same CPU object -> same GPU object returned from driver->getGPUObjectsFromAssets())
-    //use IAsset::clone()'s optional parameter to clone the buffer (being the only descriptor in default DS1) as well
-    auto ds1 = ds1_bundle.getContents().first[0];
+    uint32_t ds1UboBinding = 0u;
+    for (const auto& bnd : ds1layout->getBindings())
+        if (bnd.type==asset::EDT_UNIFORM_BUFFER)
+        {
+            ds1UboBinding = bnd.binding;
+            break;
+        }
+
+    constexpr size_t STD140_ROW_MAJOR_MVP_SZ = sizeof(float)*16ull;
+    constexpr size_t STD140_ROW_MAJOR_MV_SZ = sizeof(float)*12ull;
+    constexpr size_t STD140_ROW_MAJOR_NORMAL_MAT_SZ = sizeof(float)*12ull;
+
+    size_t neededDS1UBOsz = 0ull;
+    {
+        size_t matrixSz[asset::IPipelineMetadata::ECSI_COUNT]{};
+        matrixSz[asset::IPipelineMetadata::ECSI_WORLD_VIEW_PROJ] = STD140_ROW_MAJOR_MVP_SZ;
+        matrixSz[asset::IPipelineMetadata::ECSI_WORLD_VIEW] = STD140_ROW_MAJOR_MV_SZ;
+        matrixSz[asset::IPipelineMetadata::ECSI_WORLD_VIEW_INVERSE_TRANSPOSE] = STD140_ROW_MAJOR_NORMAL_MAT_SZ;
+        auto pipelineMetadata = static_cast<const asset::IPipelineMetadata*>(mesh_raw->getMeshBuffer(0u)->getPipeline()->getMetadata());
+        for (const auto& shdrIn : pipelineMetadata->getCommonRequiredInputs())
+            if (shdrIn.descriptorSection.type==asset::IPipelineMetadata::ShaderInput::ET_UNIFORM_BUFFER && shdrIn.descriptorSection.uniformBufferObject.set==1u && shdrIn.descriptorSection.uniformBufferObject.binding==ds1UboBinding)
+                neededDS1UBOsz = std::max(neededDS1UBOsz, shdrIn.descriptorSection.uniformBufferObject.relByteoffset+matrixSz[shdrIn.type]);
+    }
+
+    auto ds1 = core::make_smart_refctd_ptr<asset::ICPUDescriptorSet>(core::smart_refctd_ptr<asset::ICPUDescriptorSetLayout>(ds1layout));
+    for (const auto& bnd : ds1layout->getBindings())
+        if (bnd.type==asset::EDT_UNIFORM_BUFFER)
+        {
+            auto& desc = ds1->getDescriptors(bnd.binding).begin()[0];
+            auto ubo = core::make_smart_refctd_ptr<asset::ICPUBuffer>(neededDS1UBOsz);
+            desc.desc = ubo;
+            desc.buffer.offset = 0ull;
+            desc.buffer.size = neededDS1UBOsz;
+        }
+
     auto ds1_raw = static_cast<asset::ICPUDescriptorSet*>(ds1.get());
-    asset::ICPUBuffer* ubo = static_cast<asset::ICPUBuffer*>(ds1_raw->getDescriptors(0u).begin()->desc.get());
 
     auto gpuds1 = driver->getGPUObjectsFromAssets(&ds1_raw,&ds1_raw+1)->front();
-    //video::IGPUBuffer* gpuubo = gpuds1->getDescriptors()...//TODO GPU DS needs some (constant?) getter to get its descriptors
+    asset::ICPUBuffer* ubo = static_cast<asset::ICPUBuffer*>(ds1_raw->getDescriptors(0u).begin()->desc.get());
     auto gpuubo = driver->getGPUObjectsFromAssets(&ubo,&ubo+1)->front();
```

Comment:
why are you even bothering with CPU objects here, just create GPU directly.

Reply:
?

Reply:
done, no more cpu ubo